### PR TITLE
[SPARK-35671][SHUFFLE][CORE] Add support in the ESS to serve merged shuffle block meta and data to executors

### PR DESCRIPTION
**Pull request by otterc**
### What changes were proposed in this pull request?
This adds support in the ESS to serve merged shuffle block meta and data requests to executors.
This change is needed for fetching remote merged shuffle data from the remote shuffle services. This is part of push-based shuffle SPIP [SPARK-30602](https://issues.apache.org/jira/browse/SPARK-30602).

This change introduces new messages between clients and the external shuffle service:

1. `MergedBlockMetaRequest`: The client sends this to external shuffle to get the meta information for a merged block. The response to this is one of these :
  - `MergedBlockMetaSuccess` : contains request id, number of chunks, and a `ManagedBuffer` which is a `FileSegmentBuffer` backed by the merged block meta file.
  - `RpcFailure`: this is sent back to client in case of failure. This is an existing message.

2. `FetchShuffleBlockChunks`: This is similar to `FetchShuffleBlocks` message but it is to fetch merged shuffle chunks instead of blocks.

### Why are the changes needed?
These changes are needed for push-based shuffle. Refer to the SPIP in [SPARK-30602](https://issues.apache.org/jira/browse/SPARK-30602).

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Added unit tests.
The reference PR with the consolidated changes covering the complete implementation is also provided in [SPARK-30602](https://issues.apache.org/jira/browse/SPARK-30602).
We have already verified the functionality and the improved performance as documented in the SPIP doc.

Lead-authored-by: Chandni Singh chsingh@linkedin.com
Co-authored-by: Min Shen mshen@linkedin.com

### CODE REVIEW

**Comment by otterc** [comment]
@mridulm @tgravescs @attilapiros @Ngone51 @Victsm Please help review this change.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/protocol/FetchShuffleBlockChunks.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+      + Encoders.IntArrays.encodedLength(reduceIds)
+      + 4 /* encoded length of chunkIds.size() */
+      + encodedLengthOfChunkIds;
+  }
+
+  @Override
+  public void encode(ByteBuf buf) {
+    super.encode(buf);
+    Encoders.IntArrays.encode(buf, reduceIds);
+    buf.writeInt(chunkIds.length);
```
Comment (mridulm) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/protocol/FetchShuffleBlockChunks.java:99]: Add a note that even though `reduceIds.length` == `chunkIds.length`, we are explicitly setting the length in interest of forward compatibility ?

**Review by mridulm — COMMENTED** [review]
Mostly looks good to me, added a few comments.
+CC @Ngone51 to take a pass too.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java [hunk 2]
```diff
@@ -88,82 +94,125 @@ public OneForOneBlockFetcher(
...
       throw new IllegalArgumentException("Zero-sized blockIds array");
     }
-    if (!transportConf.useOldFetchProtocol() && isShuffleBlocks(blockIds)) {
+    if (!transportConf.useOldFetchProtocol() && areShuffleBlocksOrChunks(blockIds)) {
       this.blockIds = new String[blockIds.length];
-      this.message = createFetchShuffleBlocksMsgAndBuildBlockIds(appId, execId, blockIds);
+      this.message = createFetchShuffleBlocksOrChunksMsg(appId, execId, blockIds);
     } else {
       this.blockIds = blockIds;
       this.message = new OpenBlocks(appId, execId, blockIds);
```
[the remark this answers is not in the document]
Comment (otterc) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java:102]: `OpenBlocks` does support chunk blocks. Have added support in `ExternalBlockHandler` with changes in `ManagedBufferIterator` to handle the `shuffleBlock` and `shuffleChunk` differently.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java [hunk 3]
```diff
@@ -187,6 +188,39 @@ public void onFailure(Throwable e) {
...
+          @Override
+          public void onSuccess(int numChunks, ManagedBuffer buffer) {
+            logger.trace("Successfully got merged block meta for shuffleId {} reduceId {}",
+              shuffleId, reduceId);
+            listener.onSuccess(shuffleId, reduceId, new MergedBlockMeta(numChunks, buffer));
+          }
+
+          @Override
+          public void onFailure(Throwable e) {
+            logger.error("Failed while getting merged block meta", e);
```
[the remark this answers is not in the document]
Comment (otterc) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java:214]: I will remove this logging here

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java [hunk 4]
```diff
@@ -88,82 +94,125 @@ public OneForOneBlockFetcher(
...
   /** Split the shuffleBlockId and return shuffleId, mapId and reduceIds. */
   private String[] splitBlockId(String blockId) {
     String[] blockIdParts = blockId.split("_");
     // For batch block id, the format contains shuffleId, mapId, begin reduceId, end reduceId.
     // For single block id, the format contains shuffleId, mapId, educeId.
-    if (blockIdParts.length < 4 || blockIdParts.length > 5 || !blockIdParts[0].equals("shuffle")) {
+    // For single block chunk id, the format contains shuffleId, reduceId, chunkId.
+    if (blockIdParts.length < 4 || blockIdParts.length > 5 ||
+      !(blockIdParts[0].equals(SHUFFLE_BLOCK_PREFIX) ||
+        !blockIdParts[0].equals(SHUFFLE_CHUNK_PREFIX))) {
```
[the remark this answers is not in the document]
Comment (otterc) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java:215]: I made a mistake here when adding `_` to `SHUFFLE_BLOCK_PREFIX` and `SHUFFLE_CHUNK_PREFIX`. 
There are UTs with valid block Ids. 
Have added another UT for checking if an invalid chunk fails.

[the remark this answers is not in the document]
Comment (otterc) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java]: I changed this method a bit more. For a shuffle chunk, the size of parts can only be 4 so added that validation. Also expanded the test.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/OneForOneBlockFetcher.java:101]: Sigh, misread it. Thanks for clarifying.

**Review by mridulm — APPROVED** [review]
Changes look good to me.
Will give some time for @Ngone51 to also review !

+CC @attilapiros, @tgravescs

File: common/network-common/src/main/java/org/apache/spark/network/protocol/Message.java [hunk 5]
```diff
@@ -37,7 +37,8 @@
     ChunkFetchRequest(0), ChunkFetchSuccess(1), ChunkFetchFailure(2),
     RpcRequest(3), RpcResponse(4), RpcFailure(5),
     StreamRequest(6), StreamResponse(7), StreamFailure(8),
-    OneWayMessage(9), UploadStream(10), User(-1);
+    OneWayMessage(9), UploadStream(10), User(-1),
+    MergedBlockMetaRequest(11), MergedBlockMetaSuccess(12);
```
Comment (dongjoon-hyun) [inline common/network-common/src/main/java/org/apache/spark/network/protocol/Message.java:41]: nit. If you don't mind, shall we keep the order by adding this before `User(-1)`?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java [hunk 6]
```diff
@@ -361,8 +361,8 @@ public int removeBlocks(String appId, String execId, String[] blockIds) {
     return numRemovedBlocks;
   }
 
-  public Map<String, String[]> getLocalDirs(String appId, String[] execIds) {
-    return Arrays.stream(execIds)
+  public Map<String, String[]> getLocalDirs(String appId, Set<String> execIds) {
```
Comment (dongjoon-hyun) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java:364]: Do we need this function signature change?

File: common/network-common/src/main/java/org/apache/spark/network/server/RpcHandler.java [hunk 7]
```diff
@@ -129,4 +137,38 @@ public void onFailure(Throwable e) {
...
+     * @param callback Callback which should be invoked exactly once upon success or failure.
+     */
+    void receiveMergeBlockMetaReq(
+        TransportClient client,
+        MergedBlockMetaRequest mergedBlockMetaRequest,
+        MergedBlockMetaResponseCallback callback);
+  }
+
+  /**
+   * A Noop implementation of {@link MergedBlockMetaReqHandler}.
```
Comment (dongjoon-hyun) [inline common/network-common/src/main/java/org/apache/spark/network/server/RpcHandler.java:162]: Could you add more description about the current usage of this Noop implementation in this PR?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java (hunk 6 above)
Comment (otterc) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java:364]: @mridulm suggested to make this change here:
https://github.com/apache/spark/pull/32140#discussion_r645290580

This change is not necessary for push-based shuffle but it was a small change so I was fine with it.

File: common/network-common/src/main/java/org/apache/spark/network/server/RpcHandler.java (hunk 7 above)
Comment (otterc) [reply common/network-common/src/main/java/org/apache/spark/network/server/RpcHandler.java]: I have added where this Noop impl is used. PTAL. Let me know if I should describe it more.

File: common/network-common/src/main/java/org/apache/spark/network/protocol/Message.java (hunk 5 above)
Comment (otterc) [reply common/network-common/src/main/java/org/apache/spark/network/protocol/Message.java]: Done so resolving

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java (hunk 6 above)
Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java:364]: This is not a public api, is it @dongjoon-hyun ?
We could keep the signature and internally use a `Set` if it is being externally depended on.

Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockResolver.java:364]: +CC @dongjoon-hyun any thoughts on the above ? Thx

**Comment by mridulm** [comment]
Are there any other comments @Ngone51, @dongjoon-hyun ? I was planning to merge it in next couple of days if there are none.

**Comment by dongjoon-hyun** [comment]
I have no other comments, @mridulm . :)

**Comment by mridulm** [comment]
Thanks for the review @dongjoon-hyun !

**Comment by mridulm** [comment]
Merged to master, thanks for working on this @otterc.
Thanks for the review @dongjoon-hyun !

**Comment by otterc** [comment]
Thanks @mridulm and @dongjoon-hyun

**Pull request closed without merging**